ext/drmgr/drmgr.h

```diff
@@ -515,6 +515,79 @@ bool
 drmgr_unregister_opcode_instrumentation_event(drmgr_opcode_insertion_cb_t func,
                                               int opcode);
 
+DR_EXPORT
+/**
+ * Registers a callback function for the fifth instrumentation stage:
+ * post-instrumentation analysis and transformations on each
+ * basic block.  drmgr will call \p func as the fifth of five
+ * instrumentation stages for each dynamic application basic block.
+ * Post-instrumentation passes are allowed to insert both meta and
+ * non-meta instructions, and are primarily intended for debugging prior
+ * instrumentation passes.
+ *
+ * All instrumentation must follow the guidelines for
+ * #dr_register_bb_event().
+ *
+ * \return false if the given priority request cannot be satisfied
+ * (e.g., \p priority->before is already ordered after \p
+ * priority->after) or the given name is already taken.
+ *
+ * @param[in]  func        The callback to be called.
+ * @param[in]  priority    Specifies the relative ordering of the callback.
+ *                         Can be NULL, in which case a default priority is used.
+ */
+bool
+drmgr_register_bb_post_instru_event(drmgr_xform_cb_t func, drmgr_priority_t *priority);
+
+DR_EXPORT
+/**
+ * Unregisters a callback function for the fifth instrumentation stage.
+ * \return true if unregistration is successful and false if it is not
+ * (e.g., \p func was not registered).
+ *
+ * The recommendations for dr_unregister_bb_event() about when it
+ * is safe to unregister apply here as well.
+ */
+bool
+drmgr_unregister_bb_post_instru_event(drmgr_xform_cb_t func);
+
+DR_EXPORT
+/**
+ * Registers callbacks for all five instrumentation passes at once, with a \p
+ * user_data parameter passed among them all, enabling data sharing for all
+ * five of them.  See the documentation for drmgr_register_bb_app2app_event(),
+ * drmgr_register_bb_instrumentation_event(), drmgr_register_bb_instru2instru_event(), and
+ * drmgr_register_bb_post_instru_event() for further details of each pass. The
+ * aforementioned routines are identical to this with the exception of the extra \p
+ * user_data parameter, which is an OUT parameter to the \p app2app_func and passed in to
+ * the four subsequent callbacks. The \p priority param can be NULL, in which case a
+ * default priority is used.
+ */
+bool
+drmgr_register_bb_instrumentation_all_events(drmgr_app2app_ex_cb_t app2app_func,
```

Comment:
I'm skeptical about naming this "*_all_events". If we add another phase in future, it will be confusing and we wouldn't be able to rename later due to issues with binary compatibility.

Reply:
> Derek also mentioned the convention to use an _ex suffix when we later extend an API we've already released. So, there is a process to adjust for released API.

I think the _ex suffix allows us to add a new version of a function, for supporting a different signature. But the old versions still have to stay, for binary compatibility (we cannot remove the old functions or users using it in their clients will see errors). So the `drmgr_register_bb_instrumentation_all_events` with 5 callbacks cannot be removed, which can be confusing because "all events" may not always mean 5 callbacks.

Another possible way to add `drmgr_register_bb_instrumentation_all_events` is by taking as argument a struct for all callbacks. This way, even if we add a new phase in future, the struct arg will never need to change, we'll just add a new member to the end of the struct.

Reply:
+1 to using a struct: it makes backward compatibility much easier for function argument changes.

> Not sure if the struct would work 

See how we use structs throughout the API: a size field as the first field provides the versioning and we can then append fields in future versions without breaking anything.

Reply:
Right, a size field would resolve the concern I had. Changed it to use a struct.

Reply:
Resolve?

Reply:
Yep resolved. Use of the \*all_events\* is now okay given the use of the struct that allows for backward compatibility.